Dockerfile

```diff
@@ -103,13 +110,47 @@ RUN --mount=target=. --mount=target=/root/.cache,type=cache \
   xx-go build ${GOBUILDFLAGS} -ldflags "$(cat /tmp/.ldflags) -extldflags '-static'" -tags "osusergo netgo static_build seccomp ${BUILDKITD_TAGS}" -o /usr/bin/buildkitd ./cmd/buildkitd && \
   xx-verify ${VERIFYFLAGS} /usr/bin/buildkitd
 
-FROM scratch AS binaries-linux
+# build buildkitd binary (debug variant)
+FROM buildkit-base AS buildkitd-debug
+# BUILDKITD_TAGS defines additional Go build tags for compiling buildkitd
+ARG BUILDKITD_TAGS
+ARG TARGETPLATFORM
+ARG GOBUILDFLAGS
+ARG VERIFYFLAGS="--static"
+ARG CGO_ENABLED=0
+RUN --mount=target=. --mount=target=/root/.cache,type=cache \
+  --mount=target=/go/pkg/mod,type=cache \
+  --mount=source=/tmp/.ldflags,target=/tmp/.ldflags,from=buildkit-version \
+  xx-go build ${GOBUILDFLAGS} -gcflags="all=-N -l" -ldflags "$(cat /tmp/.ldflags) -extldflags '-static'" -tags "osusergo netgo static_build seccomp ${BUILDKITD_TAGS}" -o /usr/bin/buildkitd ./cmd/buildkitd && \
+  xx-verify ${VERIFYFLAGS} /usr/bin/buildkitd
+
+FROM buildkitd-${BUILD_VARIANT} AS buildkitd
+
+FROM scratch AS binaries-linux-base
 COPY --link --from=runc /usr/bin/runc /buildkit-runc
 # built from https://github.com/tonistiigi/binfmt/releases/tag/buildkit%2Fv7.1.0-30
 COPY --link --from=tonistiigi/binfmt:buildkit-v7.1.0-30@sha256:45dd57b4ba2f24e2354f71f1e4e51f073cb7a28fd848ce6f5f2a7701142a6bf0 / /
 COPY --link --from=buildctl /usr/bin/buildctl /
+
+FROM binaries-linux-base AS binaries-linux-release
 COPY --link --from=buildkitd /usr/bin/buildkitd /
 
+FROM binaries-linux-base AS binaries-linux-debug
+COPY --link --from=buildkitd /usr/bin/buildkitd /buildkitd.bin
+COPY --link --from=dlv /usr/bin/dlv /
+COPY --link --chmod=755 <<EOF /buildkitd
```

Comment:
As this is not really buildkitd but delve I think we should not call it like this and instead change entrypoint if needed (or can we just set the binary name on container start).

Reply:
I'll give it a try, but this can also get a bit complicated. I'd like to just have `ARG BUILDKITD_DEBUG=0` and to set that to 1 when I want the debug, but I have some difficulty when it comes to conditional logic.

The way the current dockerfile works is by using the name as a way to pseudo create conditionals. I can do:

```
FROM buildkit-export AS buildkit-release

FROM buildkit-export AS buildkit-debug
# debug specific customizations

FROM buildkit-${VARIANT} AS buildkit
```

This doesn't really work well with numeric values because that last line doesn't work properly. If you know of another way to do this, then I'd much prefer to make the argument a boolean.

Reply:
Found a way in the docs.

> ${variable:+word} indicates that if variable is set then word will be the result, otherwise the result is the empty string.